lightning/src/util/ser.rs

```diff
@@ -34,65 +34,49 @@ use util::byte_utils::{be64_to_array, be48_to_array, be32_to_array, be16_to_arra
 /// serialization buffer size
 pub const MAX_BUF_SIZE: usize = 64 * 1024;
 
-/// A trait that is similar to std::io::Write but has one extra function which can be used to size
+/// A struct that is similar to std::io::Write but has one extra function which can be used to size
 /// buffers being written into.
 /// An impl is provided for any type that also impls std::io::Write which simply ignores size
 /// hints.
 ///
 /// (C-not exported) as we only export serialization to/from byte arrays instead
-pub trait Writer {
+pub struct Writer(pub Vec<u8>);
+
+impl Writer {
+	/// Construct an empty Writer
+	pub fn new() -> Self {
+		Writer(Vec::new())
+	}
+
 	/// Writes the given buf out. See std::io::Write::write_all for more
-	fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error>;
+	pub fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error> {
+		self.0.write_all(buf)
+	}
 	/// Hints that data of the given size is about the be written. This may not always be called
 	/// prior to data being written and may be safely ignored.
-	fn size_hint(&mut self, size: usize);
-}
+	pub fn size_hint(&mut self, size: usize) {
+		self.0.reserve(size)
+	}
 
-impl<W: Write> Writer for W {
-	#[inline]
-	fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error> {
-		<Self as ::std::io::Write>::write_all(self, buf)
+	/// Return the number of bytes written
+	pub fn len(&self) -> usize {
+		self.0.len()
 	}
-	#[inline]
-	fn size_hint(&mut self, _size: usize) { }
 }
 
-pub(crate) struct WriterWriteAdaptor<'a, W: Writer + 'a>(pub &'a mut W);
-impl<'a, W: Writer + 'a> Write for WriterWriteAdaptor<'a, W> {
-	fn write_all(&mut self, buf: &[u8]) -> Result<(), ::std::io::Error> {
-		self.0.write_all(buf)
-	}
+pub(crate) struct WriterWriteAdaptor<'a>(pub &'a mut Writer);
+
+impl<'a> Write for WriterWriteAdaptor<'a> {
```

Comment:
Is this adaptor still needed?

Reply:
it's used in `impl_consensus_ser!` for serializing bitcoin structs

Reply:
that said, it's better if we implement `Write` directly on `Writer` - will do that now